### PR TITLE
ScalarInt: size mismatches are a bug, do not delay the panic

### PR DESCRIPTION
**Pull request by RalfJung**
Cc [Zulip](https://rust-lang.zulipchat.com/#narrow/stream/146212-t-compiler.2Fconst-eval/topic/Why.20are.20ScalarInt.20to.20iN.2FuN.20methods.20fallible.3F)

r? @oli-obk

### CODE REVIEW

**Comment by RalfJung** [comment]
> `   #11 ERROR: process "/bin/sh -c pip3 install --no-deps --no-cache-dir --require-hashes -r /tmp/reuse-requirements.txt     && pip3 install virtualenv" did not complete successfully: exit code: 1`

That looks unrelated to my PR.

File: compiler/rustc_hir_typeck/src/pat.rs [hunk 1]
```diff
@@ -2385,11 +2385,10 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         min_len: u64,
     ) -> (Option<Ty<'tcx>>, Ty<'tcx>) {
         let len = match len.eval(self.tcx, self.param_env, span) {
-            // FIXME(BoxyUwU): Assert the `Ty` is a `usize`?
```
Comment (oli-obk) [inline compiler/rustc_hir_typeck/src/pat.rs:2388]: This assert hasn't been added

Comment (RalfJung) [reply compiler/rustc_hir_typeck/src/pat.rs:2388]: `to_target_usize` is implicitly asserting that

**Pull request merged**